py-grpcio: add internal-re2 variant

viperML opened the pull request:
As explained on the comment, grpcio crashes at runtime with the error described at https://github.com/spack/spack/issues/49338.

The re2 dependency is optional so it can be safely disabled by default.

Found this when updating py-ray.
CC @theabm

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

AbhinavPraveen commented on repos/spack_repo/builtin/packages/py_grpcio/package.py:
```diff
@@ -106,24 +110,25 @@ def setup_build_environment(self, env: EnvironmentModifications) -> None:
-            r"\1('{0}',)".format(self.spec["re2"].prefix.include),
-            "setup.py",
-        )
+        with when("+re2"):
```
Is this line in the wrong place? Shouldn't it be above only the `filter_file` for `RE2_INCLUDE`?

alalazo replied:
The `when` context manager cannot be used in a method. It's meant for directives in the class scope  @viperML Did you check the package behave as intended?

viperML replied:
Oops, should had taken another coffee...

viperML committed:
py-grpcio: add internal-re2 variant

viperML commented:
@AbhinavPraveen @alalazo

Summary of the last force-push:

- Fixed the position and removed the bad `with`.
- Renamed the variant to `internal-re2`. This is probably in line with the naming conventions for other packages, i.e. `openmpi +internal-hwloc`.
- I built and ran the package with both specs. `~internal-re2` crashes with the mentioned bug, `+internal-re2` works OK, and is the new default.
- I tried pinning re2 to the same git submodule that upstream uses (`re2@2022-04-01`, https://github.com/grpc/grpc/pull/31665), but still crashes with the same error.
